### ISSUE TITLE
Consider refactoring zsh completions

### ISSUE THREAD

**Issue by DannyBen** [DannyBen/alf#41]
The discussion in #39, and the subsequent fix in #40, added this code:

https://github.com/DannyBen/alf/blob/119ffe74f47d475ead81bf635c8ca889c410bc6f/alf#L393-L405

which feels too hacky.

In addition, these calls to `compinit` and `bashcompinit` should only be called once prior to any bash `complete` calls.

Instead, we should consider the following:

1. Remove this piece of code.
2. Wrap all the `complete` calls inside `if command -v complete` so that it is executed only if `complete` is enabled.
3. Add note for zsh users to add `autoload -U +X compinit && compinit && autoload -U +X bashcompinit && bashcompinit` before sourcing `~/.bash_aliases`, if needed.

cc: @akefirad - would be interested to hear your thoughts.

**Comment by akefirad** [comment]
Good point. If this is really the case (the logic of how/when to call `compinit`), then I guess no completion script should call it. On the other hand, it's pretty hard to keep this clean (asking users to add this to their RC script, etc). Not sure what the best solution is, TBH. 🤔 (Let's see what NVM people say about this.)

**Comment by DannyBen** [comment]
Yes. Sounds like a good approach to wait for the NVM guys, assuming they will respond.

I am pretty sure the `compinit` thing should only be called once - a) it is an initialization script, and b) as we have seen, it actually breaks things if called more than once...

The truth is that I feel zsh guys complicated this aspect without any good reason, as far as I can see (or at least without consideration of the world outside zsh...). These should have been autoloaded by default to all zsh users, and be done with it.

**Comment by akefirad** [comment]
Right, seems a bit messed up. See #39 for the newest version of the code. I tested, still same issue though.

**Comment by akefirad** [comment]
Hi @DannyBen, so I did some more research and my conclusion is that `*compinit` should not be called in any completion script. The main reason, as you figured it out, is `compinit` (as well `bashcompinit`) must be called only once (and before any completion scripts). As matter of fact, this is how it's done in all other completion scripts, all asking users to add this part manually to their `.zshrc` (if not already done). The only notable exception is https://github.com/ohmyzsh/ohmyzsh which actually does a lot of magic with everything including `compinit` and (`bashcompinit`). In fact, when I disabled the part enabling `compinit` in `.bash_aliases` script generated by `alf`, everything works just fine. 🙂
More than that, if you ask users to do it (in case they don't have oh-my-zsh), since there's no need to check anything, the code is gonna be as simple as: `autoload -Uz +X compinit && compinit` (and similarly for `bashcompinit`). Also note the `-Uz` where `z` is forcing the zsh-style, recommended everywhere. WDYT?

**Comment by DannyBen** [comment]
Wow, thanks for what seems to be an extensive research.

This conclusion of yours aligns with my initial gut feeling. These "initializations" should be done by the shell (or by users of the shell) and not by any script. 

Just to be sure we are on the same page, as I understand, we should:

1. Remove the entire piece of code dealing with `compinit` and `bashcompinit` (the code mentioned at the top of the issue).
2. Add some instructions to README for zsh users.

Right?

**Comment by akefirad** [comment]
Right. Regarding the instruction, something along the lines:

Oh-My-Zsh users may skip this step! ZSH users, if haven't done yet, should add `compinit` (and `bashcompinit`) to their `.zshrc` before sourcing `.bash_aliases`:
```
autoload -Uz +X compinit && compinit
autoload -Uz +X bashcompinit && bashcompinit
```

**Comment by DannyBen** [comment]
Can you see version 0.4.9? (confirm its fine for you)

The README instructions are [here](https://github.com/DannyBen/alf#zsh-users)

**Comment by akefirad** [comment]
Works like a charm.

**Comment by DannyBen** [comment]
Excellent. 

I will go ahead and reclose this issue, feel free to comment or open a new one if problems arise.

Thanks for helping sort this out (I guess the NVM guys should do the same).
